Page loaded before all requests done

EvaSDK opened the pull request:
I was trying to compute checksum of resources downloaded by Ghost.py but some sites stream their content. This is not a problem for regular web content but file that go through `unsupported_content` do not behave properly.

One resource is created for each chunk of the file received which obvsiouly does not help in any way with getting the complete content that we can expect. The fix here is to just let `NetworkAccessManager` do its job.

The second problem that was detected with this case is that `QtWebKit` emits `pageLoaded` even though the `QNetworkReply` object in charge of downloading the file is still running. Hence keep a registry of in-flight requests and only allow `wait_for_page_loaded` to return when both `pageLoaded` has been emited and no queries are still running.

This is basically what was reported in PR #265 hopefully with a clearer wording.

Commit:
Added display server debugging messages

Signed-off-by: Gilles Dartiguelongue <g.dartiguelongue@lexfo.fr>

EvaSDK committed:
Leave NetworkAccessManager do its job

Unsupported content goes through NetworkAccessManager as well,
no need to make it special for downloading.

EvaSDK committed:
Keep a registry of in-flight requests and wait for them to complete

Most of the time, QtWebkit emits pageLoaded when all resources are
indeed loaded, however when downloading a file directly for example,
the signal is emitted even though content is still flowing down.

Keeping a registry allows delaying closing the session until all
requests created during the session are indeed complete.

EvaSDK committed:
Log download progress of responses

Some responses take a while to download so have some logs to see what is
going on. This code should probably be enhanced to skip small downloads
and or start emitting logs if downloads takes more than a pre-defined
amount of time but for now it is more helpful as is to help debug
network problems in the current code.

EvaSDK commented:
On the topic of in-flight requests tracking, see also PR #274. It is imho better to handle the registry in one place and not mix registry handling code between `NetworkAccessManager` and `Session`.

EvaSDK commented:
After more work on the topic, it appears my fix for unsupported content is wrong.

The last commit I added tried to fix normal content downloading by removing the `replyReadyRead` function because this function accumulated excessive data in `reply.data` for unsupported content.

The reason for this is that there is most likely another callback connected to this slot that reads the reply's buffer for normal content but not for unsupported one.

I'll push an update shortly.

EvaSDK committed:
Add callback to consume data for unsupported content download

EvaSDK committed:
Rename replyReadyRead callback and add docstring

The method actually calls peek, not read. A new method will be added
that uses read and does consume the reply buffer data.

EvaSDK committed:
Do not use lambda to declare signal callbacks

As written at [1], this might be a cause for the segfaults observed at
interpreter shutdown time.

[1] http://enki-editor.org/2014/08/23/Pyqt_mem_mgmt.html

EvaSDK commented:
Everything should be fine now.

EvaSDK committed:
Do not error out when receiving QNetworkReply two times

EvaSDK committed:
Log a message when a QtNetworkReply errors out

EvaSDK commented:
Would it be possible to get a feedback on this PR ? Is it of any interest to you ?

EvaSDK committed:
Log a message when the QT application is stopping

EvaSDK committed:
Handle QNetworkReply not sending required signals

Cannot get my mind around this problem so implement a workaround for now.

jeanphix committed:
Use a modern defaut user agent

Add missing Mozilla/

Signed-off-by: Gilles Dartiguelongue <g.dartiguelongue@lexfo.fr>

jeanphix committed:
Use `xvfbwrapper`

jeanphix commented:
@EvaSDK Could be nice to get this one rebased on `dev`?

EvaSDK commented:
Actually I have been trying to cleanup that branch a couple of times already but with dev now being python3 only, it is unlikely that I'll work on merging it as the production I run is still using python 2 and this is what I am targeting for the coming weeks.

jeanphix commented:
@EvaSDK As ``PySide2`` supports ``python2``, we could probably revert this choice.

EvaSDK committed:
Fix rendering of URLs in logging messages

EvaSDK committed:
Enable coverage reports from unittests run

EvaSDK committed:
Unlock xvfbwrapper dependency to compatible releases

EvaSDK committed:
Enhance display server handling log messages

EvaSDK committed:
Read version without importing the module

Keep version introspectable while avoiding ImportError when dealing with
setup.py.

EvaSDK committed:
Use isort for consistent import sorting

EvaSDK committed:
Remove outdated note about pyside_postinstall script

Closes #271. Refs #268 and #269.

EvaSDK committed:
Merge branch 'download-unsupport-content-fix' into ghost-py-0-2

Note that this seems to reveal a problem with requests being still in
flight while the page is considered loaded which might break some script
that relied on the previously broken behavior. Will fix it in an
upcoming merge request.

EvaSDK committed:
Reduce py2/3 difference when dealing with files

Also read files in binary mode as this is the expected behavior for
this kind of HTTP transfers.

EvaSDK committed:
Add fonction to deal with varying Qt bindings bytes/string representations

EvaSDK committed:
Restrict content encoding to text types

Behave more like a real browser and only care about text/* Content-Type
when reading content to encode it properly.

Other content is now intended to be available as bytes. Update unittests
to reflect this. Fixes tests under PyQt4 as well.

EvaSDK committed:
Merge branch 'content-encoding-and-pyqt4' into ghost-py-0-2

EvaSDK committed:
Remove unneeded sleep

With all signals properly connected, I could not find a reason to keep this around.

EvaSDK committed:
Fix calls to Ghost.__del__ on missing binding condition

Also change the generic Exception by the more specific RuntimeError.

EvaSDK committed:
Actually log application stop if it is started

EvaSDK committed:
Do not use sleep unnecessarily

Just call in QT event processing and avoid unneeded sleep time.

EvaSDK committed:
Increase reactivity

Reduce time spent just sleeping and allow more QT event processing to
happen according to actual time value passed to sleep and wait_for.

EvaSDK committed:
Add missing tox.ini to MANIFEST.in

EvaSDK committed:
Merge branch 'misc-fixes' into ghost-py-0-2

EvaSDK committed:
Use super() in NetworkAccessManager

Because super is super.

EvaSDK committed:
Log in-flight requests count when timeout occurs
